File: Ozan/day_1.java
// Solution for 20. Valid Parentheses
// link: https://leetcode.com/problems/valid-parentheses/description/


class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (Character ch : s.toCharArray()) {
            if (List.of('(','[','{').contains(ch)) {
                stack.push(ch);
            } else {
                if (stack.isEmpty()) 
                    return false;
                
                Character topElement = stack.peek();
                boolean validation = (topElement.equals('(') && ch.equals(')'))
                    || (topElement.equals('[') && ch.equals(']'))
                    || (topElement.equals('{') && ch.equals('}'));
                
                if (validation)
                    stack.pop();
                else
                    return false;
            }
        }

        if (stack.isEmpty())
            return true;
        return false;
    }
}

File: Murat/day_1.py
s = "{[]}"

class Solution:
    # https://leetcode.com/problems/valid-parentheses/submissions/1180235993
    def isValid(self, s: str) -> bool:
        dimensions = [0,0,0,[]]
        for i in s:
            if i == "(":
                dimensions[0] += 1
                dimensions[3].append(0)
            elif i == ")":
                if dimensions[3][-1] == [] or dimensions[3][-1] != 0:
                    return False
                dimensions[3].pop()
                dimensions[0] -= 1
            elif i == "[":
                dimensions[1] += 1
                dimensions[3].append(1)
            elif i == "]":
                if dimensions[3][-1] == [] or dimensions[3][-1] != 1:
                    return False
                dimensions[3].pop()
                dimensions[1] -= 1
            elif i == "{":
                dimensions[2] += 1
                dimensions[3].append(2)
            elif i == "}":
                if dimensions[3][-1] == [] or dimensions[3][-1] != 2:
                    return False
                dimensions[3].pop()
                dimensions[2] -= 1
            if dimensions[0] < 0 or dimensions[1] < 0 or dimensions[2] < 0:
                return False
        if dimensions[0] == 0 and dimensions[1] == 0 and dimensions[2] == 0:
            return True
        
sol = Solution()
print(sol.isValid(s))